File: examples/zh/eval/openinference/agent.py
"""真的经 LangChain 的 create_agent 跑一遍 agent loop（LangChain 1.x 的推荐
写法，内部是个 LangGraph 单节点循环），走 OpenAI 兼容接口。

这个模块会 import LangChain，所以调用方必须保证 observability.py 已经先
import 过（见 server.py 里的 import 顺序），不然 LangChainInstrumentor 挂
不上 callback，span 会静默消失。
"""

from __future__ import annotations

import os
from typing import Any

from langchain.agents import create_agent
from langchain_core.messages import AIMessage, HumanMessage, ToolMessage
from langchain_openai import ChatOpenAI

from tools import TOOLS

_AGENT_SYSTEM_PROMPT = (
    "你是一个只有两个工具的助手：get_weather 查城市天气，calculate 算算术表达式。"
    "能用工具回答的问题必须调用工具，不要凭空编数字。"
)

_agent: Any = None


def build_agent() -> Any:
    model = ChatOpenAI(
        model=os.getenv("AGENT_MODEL", "gpt-4o-mini"),
        base_url=os.getenv("OPENAI_BASE_URL") or None,
        api_key=os.getenv("OPENAI_API_KEY"),
    )
    return create_agent(model=model, tools=TOOLS, system_prompt=_AGENT_SYSTEM_PROMPT)


def get_reply(message: str) -> tuple[str, list[dict[str, Any]]]:
    global _agent
    if _agent is None:
        if not os.getenv("OPENAI_API_KEY"):
            raise RuntimeError("缺少 OPENAI_API_KEY：在 .env 里配置真实的模型访问凭证后重启。")
        _agent = build_agent()

    result = _agent.invoke({"messages": [HumanMessage(content=message)]})
    messages = result["messages"]

    tool_calls_by_id: dict[str, dict[str, Any]] = {}
    for m in messages:
        if isinstance(m, AIMessage):
            for call in m.tool_calls or []:
                tool_calls_by_id[call["id"]] = {
                    "name": call["name"],
                    "input": call["args"],
                    "output": None,
                }
        elif isinstance(m, ToolMessage) and m.tool_call_id in tool_calls_by_id:
            tool_calls_by_id[m.tool_call_id]["output"] = m.content

    reply = next(
        (m.content for m in reversed(messages) if isinstance(m, AIMessage) and m.content),
        "",
    )
    return reply, list(tool_calls_by_id.values())

File: examples/zh/eval/langgraph/agent/agent.ts
// 真的经 LangGraph 的 createReactAgent 跑一遍 ReAct 循环
// (LLM -> 决定要不要调工具 -> 执行工具 -> 把结果喂回 LLM -> ...)。
// 这条路径会经过 @langchain/core 的埋点,配合 ../observability.ts 里注册的
// LangSmith OTel exporter 出 span。
import { AIMessage, ToolMessage, type BaseMessage } from "@langchain/core/messages";
import { MemorySaver } from "@langchain/langgraph";
import { createReactAgent } from "@langchain/langgraph/prebuilt";
import { ChatOpenAI } from "@langchain/openai";
import { agentTools } from "./tools.ts";
import type { ChatTurnResult, ToolCallRecord } from "./types.ts";

const SYSTEM_PROMPT = `你是一个乐于助人的中文 AI 助手。
需要天气信息时调用 get_weather,并用工具返回的数据作答,不要凭空编造天气。
需要精确计算时调用 calculate,把表达式交给它算,不要心算。
普通闲聊不要调用任何工具。回复保持中文、友好、简洁。`;

let cachedAgent: ReturnType<typeof buildAgent> | undefined;

function buildAgent() {
  const llm = new ChatOpenAI({
    model: process.env.AGENT_MODEL ?? "gpt-4o-mini",
    apiKey: process.env.OPENAI_API_KEY,
    configuration: process.env.OPENAI_BASE_URL ? { baseURL: process.env.OPENAI_BASE_URL } : undefined,
  });
  return createReactAgent({
    llm,
    tools: agentTools,
    prompt: SYSTEM_PROMPT,
    // 内存态 checkpointer:同一个 sessionId(-> thread_id)内的多轮对话有记忆,
    // 进程重启就丢——演示用足够了,生产场景换持久化 checkpointer。
    checkpointer: new MemorySaver(),
  });
}

function getAgent() {
  cachedAgent ??= buildAgent();
  return cachedAgent;
}

export async function runTurn(message: string, sessionId: string): Promise<ChatTurnResult> {
  const agent = getAgent();
  const result = await agent.invoke(
    { messages: [{ role: "user", content: message }] },
    { configurable: { thread_id: sessionId } },
  );
  const messages = result.messages as BaseMessage[];
  const toolCalls = extractToolCalls(messages);
  const last = messages.at(-1);
  const reply = typeof last?.content === "string" ? last.content : JSON.stringify(last?.content ?? "");
  return { reply, toolCalls };
}

/** 把这一轮里产生的 AIMessage.tool_calls + 对应 ToolMessage 配对成 {name, input, output}。 */
function extractToolCalls(messages: BaseMessage[]): ToolCallRecord[] {
  const callsById = new Map<string, { name: string; input: unknown }>();
  for (const msg of messages) {
    if (msg instanceof AIMessage) {
      for (const call of msg.tool_calls ?? []) {
        if (call.id) callsById.set(call.id, { name: call.name, input: call.args });
      }
    }
  }
  const toolCalls: ToolCallRecord[] = [];
  for (const msg of messages) {
    if (msg instanceof ToolMessage) {
      const meta = msg.tool_call_id ? callsById.get(msg.tool_call_id) : undefined;
      toolCalls.push({
        name: meta?.name ?? String(msg.name ?? "unknown_tool"),
        input: meta?.input ?? {},
        output: parseToolContent(msg.content),
      });
    }
  }
  return toolCalls;
}

function parseToolContent(content: BaseMessage["content"]): unknown {
  if (typeof content !== "string") return content;
  try {
    return JSON.parse(content);
  } catch {
    return content;
  }
}
